test/e2e/apicast_test.go

```diff
@@ -0,0 +1,152 @@
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/3scale/apicast-operator/test/e2e/e2eutil"
+
+	"github.com/3scale/apicast-operator/pkg/apis"
+	appsgroup "github.com/3scale/apicast-operator/pkg/apis/apps"
+	appsv1alpha1 "github.com/3scale/apicast-operator/pkg/apis/apps/v1alpha1"
+	framework "github.com/operator-framework/operator-sdk/pkg/test"
+	frameworke2eutil "github.com/operator-framework/operator-sdk/pkg/test/e2eutil"
+	v1 "k8s.io/api/core/v1"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	APIcastEmbeddedConfigurationSecretName = "apicast-embedded-configuration"
+)
+
+func TestAPIcastBasicDeployment(t *testing.T) {
+	err := registerAPIcastTypeInTestFramework()
+	if err != nil {
+		t.Fatalf(err.Error())
+	}
+
+	ctx := framework.NewTestCtx(t)
+	defer ctx.Cleanup()
+
+	namespace, err := ctx.GetNamespace()
+	if err != nil {
+		t.Fatal(err)
+	}
+	f := framework.Global
+	operatorName := "apicast-operator"
+	err = initializeClusterResources(t, ctx, f, namespace, operatorName)
+	if err != nil {
+		t.Fatalf("failed to initialize cluster resources: %v", err)
+	}
+	t.Log("initialized cluster resources")
+
+	apicastName := "example-apicast"
+	apicast := &appsv1alpha1.APIcast{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "example-apicast",
+			Namespace: namespace,
+		},
+		Spec: appsv1alpha1.APIcastSpec{
+			EmbeddedConfigurationSecretRef: &v1.LocalObjectReference{
+				Name: APIcastEmbeddedConfigurationSecretName,
+			},
+		},
+	}
+
+	var start time.Time
+	var elapsed time.Duration
+
+	start = time.Now()
+
+	err = f.Client.Create(context.TODO(), apicast, &framework.CleanupOptions{TestContext: ctx, Timeout: 5 * time.Minute, RetryInterval: retryInterval})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	elapsed = time.Since(start)
+
+	err = createAPIcastEmbeddedConfigurationSecret(t, ctx, f, namespace)
```

Comment:
create the secret before deploying apicast CR? You do not need to implement waiting for secret to be available, apicast deployment resource will do that for you, if I am not wrong.

Reply:
You are correct. You don't need to create the secret before deploying APIcast as the operator would return errors and requeue the event until it finds it.

That said, I don't see a problem with precreating the secret and waiting for it so we define an order. We just make sure in this test that order events happen.

Reply:
My point is, if deployment waits, why implement it?

IIRC when deployment has env var from secret and secret does not start, deployment will wait until that secret exist (doing internal polling I assume). Just for simplicity. 

Up to you